### ISSUE TITLE
resourceProjectPolicy: RESOURCE_PROJECT not respected for all API calls

### ISSUE THREAD

**Issue by tomasaschan** [GoogleCloudPlatform/k8s-config-connector#740]
### Checklist

- [X] I did not find a related open issue.
- [X] I did not find a solution in the troubleshooting guide: (https://cloud.google.com/config-connector/docs/troubleshooting)
- [X] If this issue is time-sensitive, I have submitted a corresponding issue with [GCP support](https://cloud.google.com/support-hub).

### Bug Description

Despite heavy quota bumps, our KCC setup is being rate limited on _Read request per minute_  and _Write request per minute_ quotas under `cloudresourcemanager.googleapis.com`. Audit logs show _much_ lower usage (peaks at a dozen requests per minute) for the project in question, so these requests are mostly related to resources in other projects.

![image](https://user-images.githubusercontent.com/1550920/203835071-d7bd9250-394a-4f27-9c97-370e4049b92c.png)

We run KCC in namespaced mode, with a separate service account configured for each namespaced operator deployment. These service accounts all live in a single GCP project, which is now capping out on these quotas. KCC is configured to set `resourceProjectPolicy: RESOURCE_PROJECT` everywhere, and for most quotas it works, but for some subset of requests this config value is not propagated correctly, resulting in quota usage being attributed to the project of the service account, instead of the project of the resource. With these request families for the entire organization gathered into a single project, we of course hit our quota limits quite rapidly.

### Additional Diagnostic Information

When looking at audit logs for write requests across the organization, the total number of requests per day (matching `cloudresourcemanager.googleapis.com`) has been hovering around approximately the rate limit for the past month, and we've also been maxing out the quota more or less constantly during that period. This tells me that rejected requests don't show up in our logs.

Because of the number of service accounts in the project, the metrics page (e.g. https://console.cloud.google.com/apis/api/cloudresourcemanager.googleapis.com/metrics) fails to load.

In summary, we have no observability into rejected request volumes, more specifically what type of requests they are, etc.

### Kubernetes Cluster Version

1.21 and 1.24 (and all versions in beteween) affected

### Config Connector Version

1.95

### Config Connector Mode

namespaced mode (default)

### Steps to reproduce the issue

1. Deploy KCC running as a service account X in project A, and with `resourceProjectPolicy: RESOURCE_PROJECT` set.
2. Grant access to X to also manage resources in project B.
3. Create a relevant resource for KCC to reconcile. Note that we don't know precisely which resource that causes these calls, but we think it's likely that e.g. a `IAMPartialPolicy` referencing project B in `spec.resourceRef` is one example.
4. Observe quota usage being incorrectly attributed to project A instead of B.

**Comment by diviner524** [comment]
@tomasaschan Usually the quota under cloudresourcemanager.googleapis.com is attributed to KCC resources under "Resource Manger", for example the `Project` resource kind:

https://cloud.google.com/config-connector/docs/reference/resource-docs/resourcemanager/project

There might be two potential issues causing resourceProjectPolicy: RESOURCE_PROJECT not respected:

1. You may need to use BILLING_PROJECT instead of RESOURCE_PROJECT when specifying `resourceProjectPolicy`, as it seems the underlying [terraform provider](https://github.com/hashicorp/terraform-provider-google-beta/blob/main/google-beta/resource_google_project.go#L645) only looks for BILLING_PROJECT. However note the implementation I shared is not counted towards the `cloudresourcemanager.googleapis.com` quota, instead it is for service usage (the `serviceuage.googleapis.com` quota) API calls happened during creation of a project.

2. We will need to confirm with terraform experts, but it seems the terraform google provider code is not really applying any `resourceProjectPolicy` when [GET/read a project](https://github.com/hashicorp/terraform-provider-google-beta/blob/main/google-beta/resource_google_project.go#L595). It is possible the resources under `cloudresourcemanager.googleapis.com` do not support user project overrides.

A possible future solution is to introduce a configurable reconciliation feature in KCC, that way as a KCC user you can configure the reconciliation to happen much less aggressively for desired KCC resources and avoid hitting the quota. We are currently working on this new feature and expecting to have this implemented in Q1 2023.

**Comment by tomasaschan** [comment]
Thanks for getting back! I'll address your points one by one.

> Usually the quota under cloudresourcemanager.googleapis.com is attributed to KCC resources under "Resource Manger", for example the Project resource kind

I believe the majority of the requests here, if not all, come from setting IAM policy bindings on the project, rather than managing the project itself. We don't manage [`Project`](https://cloud.google.com/config-connector/docs/reference/resource-docs/resourcemanager/project)s (or any other resources under that API group) declaratively, so there are exactly 0 requests from those resources; however, there are lots of `IAMPartialPolicy` and `IAMPolicyMember` resources which have a `Project` in their `.spec.resourceRef`. This is where I believe the quota usage is coming from.

> You may need to use BILLING_PROJECT instead of RESOURCE_PROJECT when specifying resourceProjectPolicy, as it seems the underlying [terraform provider](https://github.com/hashicorp/terraform-provider-google-beta/blob/main/google-beta/resource_google_project.go#L645) only looks for BILLING_PROJECT.

Hm, that might work - but only out of coincidence, since we also happen to isolate `ConfigConnectorContext`s per target project. (It's also a very annoying solution, since it means we have to change almost 3000 config connector contexts across several clusters, and make sure KCC doesn't overload anything while it restarts the reconcilers...) I will do some work to try to test the behavior here, and see if this can be a viable workaround for us for now.

I would argue that we shouldn't _have to_, though; `RESOURCE_PROJECT` is intended, and documented, to mean the same thing as `BILLING_PROJECT: {project the resource belongs to}`, so if that doesn't work for these resources, that's still a bug. Maybe the KCC code that calls the Terraform code needs to set `billingProject` in those parameters to the resource project, if `RESOURCE_PROJECT` is specified?

> it seems the terraform google provider code is not really applying any resourceProjectPolicy when [GET/read a project](https://github.com/hashicorp/terraform-provider-google-beta/blob/main/google-beta/resource_google_project.go#L595). It is possible the resources under `cloudresourcemanager.googleapis.com` do not support user project overrides.

Not sure if or how that statement applies to IAM bindings on the project as well, but either way I think that if this turns out to be true, it's (also) a bug in KCC, given how KCC documents how it will assign quota usage. If fixing KCC requires first building support in the underlying terraform provider, that makes it harder to fix, but it doesn't make it less of a bug 🙂

**Comment by PatrickChenSe** [comment]
> Maybe the KCC code that calls the Terraform code needs to set billingProject in those parameters to the resource project, if RESOURCE_PROJECT is specified?

Second this suggestion. I believe it's easier for KCC to verify if this works before we push changes to our production resources.

**Comment by tomasaschan** [comment]
> BILLING_PROJECT should not be seen as the "project the resource belongs to".

@diviner524 Sorry if I was unclear, but that's not at all what I said 😄 

Let's take a concrete example, to make it less ambiguous what I mean. For simplicity, consider just these two resources - a `ConfigConnectorContext` in a namespace, and an `IAMPartialPolicy` referencing the target project. These are representative for our configuration today:

```yaml
apiVersion: core.cnrm.cloud.google.com/v1beta1
kind: ConfigConnectorContext
metadata:
  name: configconnectorcontext.core.cnrm.cloud.google.com
  namespace: foo-bar
spec:
  googleServiceAccount: foo-bar@spotify-declarative-infra-mgmt.iam.gserviceaccount.com
  requestProjectPolicy: RESOURCE_PROJECT
---
apiVersion: iam.cnrm.cloud.google.com/v1beta1
kind: IAMPartialPolicy
metadata:
  namespace: foo-bar
  name: project-level-policy
spec:
  resourceRef:
    kind: Project
    external: projects/foo-bar
  bindings:
    role: roles/storage.admin
    members:
    - member: user:tomasl@spotify.com
```

What I tried to say in my previous comment, was that translating this to use `BILLING_PROJECT` instead, the docstring for `resourceProjectPolicy` states that the config above should behave equivalent to setting:

```yaml
requestProjectPolicy: BILLING_PROJECT
billingProject: foo-bar # note that we now have to specify this explicitly, to point to the project where the resource is
```

However, instead, it behaves like this, which I claim is a bug in KCC:
```yaml
requestProjectPolicy: BILLING_PROJECT
billingProject: spotify-declarative-infra-mgmt # this is incorrect!
```

Another equivalent configuration to current behavior (which might be more accurately representing _why_ this happens, as this is the default behavior):
```yaml
requestProjectPolicy: SERVICE_ACCOUNT_PROJECT
```

My claim, in summary, is that if the Terraform provider is not aware of the `RESOURCE_PROJECT` version, KCC needs to ensure that it passes the correct billing project anyway when calling out to the provider code.

> How are these Project being referenced in the IAM resources? As you mentioned you are not managing the Project declaratively using CRD, does that mean you are using spec.resourceRef.external to specify the project?

Exactly (as exemplified above).

**Comment by diviner524** [comment]
@tomasaschan Thanks for the clarification! This seems to be related to how the `google_project_iam` Terraform resource is implemented. We will do some further investigation with Terraform Google Provider team and update here with our findings.

**Comment by diviner524** [comment]
@tomasaschan

- We have confirmed with Terraform Google Provider team that `RESOURCE_PROJECT` (which is called [UserProjectOverride](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/d4f73950a07141a50385b48e13ebce7a9da4e342/pkg/controller/kccmanager/kccmanager.go#L81) in Terraform configuration) is currently not supported for [`google_project_iam`](https://registry.terraform.io/providers/hashicorp/google-beta/latest/docs/resources/google_project_iam) resource. This is the underlying Terraform resource being used in the IAMPartialPolicy example you shared earlier. This explains why `RESOURCE_PROJECT` is not respected in this case. There is a [GitHub issue](https://github.com/hashicorp/terraform-provider-google/issues/9437 ) on Terraform Google Provider repo, which tracks the missing feature, you can leave comment in that issue to help TPG team prioritize the fix. We are also working with Terraform Google Provider team internally to see if they can raise the priority for the fix of this particular resource.

-  We also tested and confirmed that "BILLING_PROJECT" will work in this case. Which means if you specify:

```yaml
requestProjectPolicy: BILLING_PROJECT
billingProject: foo-bar 
```

The API requests (`google.cloudresourcemanager.v1.Projects.GetIamPolicyGetIAMPolicy` and `
google.cloudresourcemanager.v1.Projects.SetIamPolicy`) will be attributed correctly to project `foo-bar`.

This means changing your ConfigConectorContext to use `BILLING_PROJECT` should be able to reduce the traffic volume against `spotify-declarative-infra-mgmt` project.

- Finally IIUC, you are also suggesting Config Connector to create an abstraction on top of the two features ("RESOURCE_PROJECT" and "BILLING_PROJECT") and simplify the configuration. This is definitely an interesting thought, we will created a tracking item internally to discuss the idea. But do note these two features, although sharing some similarities, still serve slightly different purposes. As an example, you can have resource `foo` from `project_foo` and resource `bar` from `project_bar` in the same K8s namespace. In this setup, specifying "RESOURCE_PROJECT" will attribute their API requests to their own "container" projects respectively. While specifying `billingProject: baz` will attribute all API requests to a single project `baz`. We will need more thoughts on these differences when designing a possible new feature.

**Comment by tomasaschan** [comment]
> We have confirmed [...]

Yeah, this is what I've been suspecting. Thanks for digging up the specifics, and confirming this is indeed the case.

> Finally IIUC, you are also suggesting Config Connector to create an abstraction on top of the two features ("RESOURCE_PROJECT" and "BILLING_PROJECT")

No, not exactly. I'm suggesting that for as long as the Terraform provider for these resources does not support UserProjectOverride, but does support explicitly setting the billing project, the KCC operator should work around this so that it behaves as documented. Since it's already possible to tell the Terraform provider what project to use as billing project, and KCC knows what project I want to use as billing project in each case, it should be able to resolve that for me without changes in our config. 

In other words, while I see how adding support in the Terraform provider makes it easier to make KCC behave as documented, I don't think a bug fix should need to be blocked by changes in the Terraform provider. 

> We also tested and confirmed that "BILLING_PROJECT" will work in this case.  

Thanks; I'll bring this up with my team. We might consider scripting this change across all those configs to get rid of our rate limiting woes.
